fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java

```diff
@@ -705,7 +729,8 @@ public CompletableFuture<Void> buildIndexAsync() {
     @VisibleForTesting
     @Nonnull
     CompletableFuture<Void> buildIndexAsync(boolean markReadable) {
-        return indexingLauncher(() -> getIndexer().buildIndexAsync(markReadable));
+        boolean useSyncLock = (!indexingPolicy.isMutual() || fallbackToRecordsScan) && common.shouldUseSynchronizedSession();
```

Comment:
At any point, if all the indexes become readable that an "all done" event. 
Assume the non-crash path: 
One thread marks all indexes as readable (in multiple transaction, because we do not want to fail all for one "uniqueness violation" or other issue). 
For each other thread, there is a very small chance of checking the indexes' state during the "mark readable" operation, hence get a false "few are readable". In this case, it will try to lock and mark the rest readable - and probably find out that they already are. 
Crash path: 
If the marking readable process fails (uniqueness violation? other crashes?), at least one other thread will try to clean up and finish marking readable.